### PR TITLE
Add timeouts for tcp syn and tcp ack

### PR DESCRIPTION
**Pull request by Gmacem**
Add syn and ack tcp timeouts for firewall

### CODE REVIEW

**Comment by taitov** [comment]
Hello! Resolve conflicts and use clang-format, please.

File: dataplane/worker_gc.cpp [hunk 1]
```diff
@@ -940,6 +949,17 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
...
 
+inline uint16_t worker_gc_t::get_acl_tcp_timeout(const uint8_t flags, const uint32_t ack_timeout, 
+																								 const uint32_t syn_timeout, const uint32_t fin_timeout) {
+	if (flags & (RTE_TCP_ACK_FLAG | RTE_TCP_SYN_FLAG)) {
+		return ack_timeout;
+	}
+	if (flags & RTE_TCP_SYN_FLAG) {
+		return syn_timeout;
+	}
+	return fin_timeout;
```
Comment (3Hren) [inline dataplane/worker_gc.cpp:960]: Seems strange to see FIN timeout as a "default" value. Maybe it's reasonable to create а real default timeout, while processing FIN timeout explicitly using mask?

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: dataplane/worker_gc.cpp [hunk 2]
```diff
@@ -931,6 +941,23 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
 	return ((uint16_t)(current_time - timestamp) > timeout);
 }
 
+inline uint16_t worker_gc_t::get_acl_tcp_timeout(const uint8_t flags, const uint32_t ack_timeout, const uint32_t syn_timeout, const uint32_t fin_timeout, const uint32_t default_timeout)
+{
+	if (flags & (RTE_TCP_ACK_FLAG | RTE_TCP_SYN_FLAG))
```
Comment (3Hren) [inline dataplane/worker_gc.cpp:946]: This condition will be `true` if any of SYN|ACK flags are set ...

File: dataplane/worker_gc.cpp [hunk 3]
```diff
@@ -931,6 +941,23 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
...
+	{
+		return ack_timeout;
+	}
+	if (flags & RTE_TCP_SYN_FLAG)
```
Comment (3Hren) [inline dataplane/worker_gc.cpp:950]: ... that means that this condition is unreachable.

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: dataplane/worker_gc.h [hunk 4]
```diff
@@ -39,6 +39,7 @@ class worker_gc_t
 	void handle_free_mbuf();
 
 	bool is_timeout(const uint16_t timestamp, const uint16_t timeout);
+	uint16_t get_acl_tcp_timeout(const uint8_t flags, const uint32_t ack_timeout, const uint32_t syn_timeout, const uint32_t fin_timeout, const uint32_t default_timeout);
```
Comment (3Hren) [inline dataplane/worker_gc.h:42]: Well, here we have timeouts' arguments type as `u32`, but return `u16` instead.

Comment (3Hren) [inline dataplane/worker_gc.h:42]: Having function with 5 arguments seems alien. I'd rather obtained concrete timeouts inside the function using `globalbase_atomic` directly.

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: dataplane/worker.cpp [hunk 5]
```diff
@@ -4008,6 +4008,12 @@ inline void cWorker::balancer_handle()
 		dataplane::globalBase::balancer_state_value_t* value;
 		dataplane::spinlock_nonrecursive_t* locker;
 		const uint32_t hash = basePermanently.globalBaseAtomic->balancer_state->lookup(key, value, locker);
+
+		rte_tcp_hdr* tcpHeader = rte_pktmbuf_mtod_offset(mbuf, rte_tcp_hdr*, metadata->transport_headerOffset);
+		if (metadata->transport_headerType == IPPROTO_TCP) {
+			value->tcp_flags = tcpHeader->tcp_flags;
```
Comment (taitov) [inline dataplane/worker.cpp:4014]: `value` maybe `nullptr`

**Review by taitov — CHANGES_REQUESTED** [review]

File: dataplane/worker_gc.cpp [hunk 6]
```diff
@@ -931,6 +938,40 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
 	return ((uint16_t)(current_time - timestamp) > timeout);
 }
 
+inline uint32_t worker_gc_t::get_acl_tcp_timeout(const uint8_t flags, const dataplane::globalBase::fw_state_config_t& fw_state_config)
+{
+	if ((flags & RTE_TCP_ACK_FLAG) && (flags & RTE_TCP_SYN_FLAG))
```
Comment (3Hren) [inline dataplane/worker_gc.cpp:943]: I'm not sure this compiles the right way without unnecessary branches.
Consider more straightforward:

```c++
if ((flags & (RTE_TCP_SYN_FLAG | RTE_TCP_ACK_FLAG)) == (RTE_TCP_SYN_FLAG | RTE_TCP_ACK_FLAG)) { ...
```

Or, as you check for SYN_FLAG the next branch bellow, you can unite them both:

```c++
if ((flags & RTE_TCP_SYN_FLAG) == RTE_TCP_SYN_FLAG) {
    if ((flags & RTE_TCP_ACK_FLAG) == RTE_TCP_ACK_FLAG) {
        return fw_state_config.tcp_ack_timeout;
    } else {
        return fw_state_config.tcp_syn_timeout;
    }
}
```

File: dataplane/worker_gc.cpp [hunk 7]
```diff
@@ -931,6 +938,40 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
...
+	{
+		return fw_state_config.tcp_ack_timeout;
```
Comment (3Hren) [inline dataplane/worker_gc.cpp:945]: Following the condition above this should be named as `tcp_syn_ack_timeout`, isn't it?

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: dataplane/type.h [hunk 8]
```diff
@@ -798,6 +802,15 @@ struct balancer_state_value_t
 	uint16_t timestamp_last_packet;
 	uint16_t timestamp_gc;
 	uint32_t nap0; ///< @todo: DELETE
+	uint8_t tcp_flags;
+};
```
Comment (taitov) [inline dataplane/type.h:806]: You can pack struct like this:
```
struct balancer_state_value_t
{
	uint32_t real_unordered_id;
	uint32_t timestamp_create; ///< @todo: 16bit
	uint16_t timestamp_last_packet;
	uint16_t timestamp_gc;
	uint32_t tcp_flags; ///< 8bit actually
};
```
or:
```
struct balancer_state_value_t
{
	uint32_t real_unordered_id;
	uint16_t tcp_flags; ///< 8bit actually
	uint16_t timestamp_create;
	uint16_t timestamp_last_packet;
	uint16_t timestamp_gc;
	uint32_t nap0; ///< @todo: DELETE
};
```

File: dataplane/worker_gc.cpp [hunk 9]
```diff
@@ -484,7 +484,11 @@ void worker_gc_t::handle_balancer_gc()
 				iter.lock();
 			}
 
-			if ((uint16_t)((uint16_t)globalbase_atomic->currentTime - (uint16_t)iter.value()->timestamp_last_packet) > balancer_state_ttl)
+			auto tcp_timeout = get_balancer_tcp_timeout(
+			        iter.value()->tcp_flags,
+			        globalbase_atomic->balancer_state_config);
+			// if ((uint16_t)((uint16_t)globalbase_atomic->currentTime - (uint16_t)iter.value()->timestamp_last_packet) > balancer_state_ttl) {
```
[the remark this answers is not in the document]
Comment (taitov) [reply dataplane/worker_gc.cpp:490]: `currentTime` - is 32bit, `timestamp_last_packet` - 16bit.
The result of subtracting these values will always be greater than 60.

**Review by 3Hren — APPROVED** [review]
LGTM

**Comment by 3Hren** [comment]
Still, we have u32 -> u16 implicit conversions. Delegate further review to @taitov.

File: dataplane/dataplane.cpp [hunk 10]
```diff
@@ -91,6 +91,15 @@ cDataPlane::cDataPlane() :
 	                {eConfigType::master_mempool_size, 8192},
 	                {eConfigType::nat64stateful_states_size, YANET_CONFIG_NAT64STATEFUL_HT_SIZE},
 	                {eConfigType::kernel_interface_queue_size, YANET_CONFIG_KERNEL_INTERFACE_QUEUE_SIZE},
+	                {eConfigType::balancer_state_ttl, YANET_CONFIG_BALANCER_STATE_TIMEOUT_DEFAULT},
+	                {eConfigType::balancer_state_ht_size, YANET_CONFIG_BALANCER_STATE_HT_SIZE},
```
Comment (taitov) [inline dataplane/dataplane.cpp:95]: Duplicate `balancer_state_ttl` and `balancer_state_ht_size`

File: dataplane/dataplane.cpp [hunk 11]
```diff
@@ -1945,6 +1954,33 @@ eResult cDataPlane::parseConfigValues(const nlohmann::json& json)
...
+	{
+		configValues[eConfigType::balancer_tcp_syn_timeout] = json["balancer_tcp_syn_timeout"];
+	}
+	if (exist(json, "balancer_tcp_fin_timeout"))
+	{
+		configValues[eConfigType::balancer_tcp_fin_timeout] = json["balancer_tcp_fin_timeout"];
+	}
+	if (exist(json, "balancer_tcp_timeout"))
+	{
+		configValues[eConfigType::balancer_tcp_fin_timeout] = json["balancer_tcp_timeout"];
```
Comment (taitov) [inline dataplane/dataplane.cpp:1983]: `}` ?

File: dataplane/worker_gc.cpp [hunk 12]
```diff
@@ -484,7 +484,10 @@ void worker_gc_t::handle_balancer_gc()
 				iter.lock();
 			}
 
-			if ((uint16_t)((uint16_t)globalbase_atomic->currentTime - (uint16_t)iter.value()->timestamp_last_packet) > balancer_state_ttl)
+			auto tcp_timeout = get_balancer_tcp_timeout(
+			        iter.value()->tcp_flags,
+			        globalbase_atomic->balancer_state_config);
+			if ((uint16_t)((uint16_t)globalbase_atomic->currentTime - (uint16_t)iter.value()->timestamp_last_packet) > tcp_timeout)
```
Comment (taitov) [inline dataplane/worker_gc.cpp:490]: For 16bit timestamps, you can use `is_timeout()`

File: dataplane/worker_gc.cpp [hunk 13]
```diff
@@ -931,6 +937,40 @@ inline bool worker_gc_t::is_timeout(const uint16_t timestamp,
...
+		return fw_state_config.tcp_syn_timeout;
+	}
+	if (flags & RTE_TCP_FIN_FLAG)
+	{
+		return fw_state_config.tcp_fin_timeout;
+	}
+	return fw_state_config.tcp_timeout;
+}
+
+inline uint32_t worker_gc_t::get_balancer_tcp_timeout(const uint8_t flags, const dataplane::globalBase::balancer_state_config_t& balancer_state_config)
```
Comment (taitov) [inline dataplane/worker_gc.cpp:957]: ... and write one function with `const state_config_t& state_config`

**Review by taitov — CHANGES_REQUESTED** [review]

**Review by GeorgyKirichenko — CHANGES_REQUESTED** [review]

**Review by GeorgyKirichenko — DISMISSED** [review]
LGTM

File: dataplane/type.h [hunk 14]
```diff
@@ -794,11 +796,23 @@ static_assert(YANET_CONFIG_BALANCERS_SIZE <= 0x7F, "invalid size");
 struct balancer_state_value_t
 {
 	uint32_t real_unordered_id;
-	uint32_t timestamp_create; ///< @todo: 16bit
+	uint32_t state_timeout;
+	uint32_t nap0; ///< @todo: DELETE
+	uint16_t timestamp_create;
 	uint16_t timestamp_last_packet;
 	uint16_t timestamp_gc;
-	uint32_t nap0; ///< @todo: DELETE
```
Comment (taitov) [inline dataplane/type.h:800]: Then just rename `nap0` to `state_timeout`

File: dataplane/worker_gc.cpp [hunk 15]
```diff
@@ -483,12 +483,14 @@ void worker_gc_t::handle_balancer_gc()
 
 				iter.lock();
 			}
-
-			if ((uint16_t)((uint16_t)globalbase_atomic->currentTime - (uint16_t)iter.value()->timestamp_last_packet) > balancer_state_ttl)
 			{
-				const auto& real_from_base = base.globalBase->balancer_reals[iter.value()->real_unordered_id];
-				++counters[real_from_base.counter_id + (tCounterId)balancer::gc_real_counter::sessions_destroyed];
-				iter.unset_valid();
+				auto value = *iter.value();
```
Comment (taitov) [inline dataplane/worker_gc.cpp:487]: Why you copy `value` in this place? Maybe use pointer `auto* value = iter.value()`?

**Review by taitov — APPROVED** [review]

**Pull request merged**